### ISSUE TITLE
Screen: on_enter, on_leave default handlers not called when callback is bound to event

### ISSUE THREAD

**Issue by rafalo1333** [kivy/kivy#4260]
When on_enter/on_leave (possibly on_pre_* also) handler methods are declared in subclass of Screen, they are not called in instances when there is at least one callback bound to its corresponding events.

https://gist.github.com/rafalo1333/cbe4d32d2faa2a1200404a0273ee88f3

**Comment by bionoid** [comment]
The problem appears to be that ``Popup.dismiss`` returns ``self``. One workaround is to bind to something that returns a false value

```python
screen2.bind(on_enter=lambda x: p.dismiss() and False)
```

**Comment by rafalo1333** [comment]
Oh, that is interesting... I believe it is not great that kivy core widgets used as docs state could prevent various events from firing it's handlers? Is it common to all Kivy events?

**Comment by dessant** [comment]
Yes, events stop propagating when a callback returns True.

**Comment by dessant** [comment]
> I believe it is not great that kivy core widgets used as docs state could prevent various events from firing it's handlers

Have you taken the on_enter bind from the docs?

**Comment by kived** [comment]
We could document this better, like in a more general section on the "Events and Properties" page. Right now it's buried in a specific example about Properties and a section on the Widget page about touch events.

https://kivy.org/docs/guide/events.html#dispatching-a-property-event
https://kivy.org/docs/api-kivy.uix.widget.html#widget-event-bubbling

**Comment by rafalo1333** [comment]
I believe widget.bind(event=callback) is rather standard way of binding in Kivy... I was not mentioning individual docstring, rather the standard. I am surprised that few widgets' events are able to block another callbacks. I wasnt expecting the Popup to do such a harm in my code.

Maybe it needs a doc improvement for EventDispatcher? There is some info (about callback returning True makes event unsuccessful) but we need some examples, the Popup is first I am thinging about.

**Comment by kived** [comment]
Returning `True` (or any truthy value) marks the event as "handled" and it will no longer propagate. This is a common feature of most event systems. For example, Javascript does the same thing (though opposite logic, the return value specifies whether the event should continue propagating rather than whether the event has been handled).

**Comment by rafalo1333** [comment]
I am not saying it's wrong (as I am still learning new things) but I believe that if its better documented, it can be less shocking :)

**Comment by dessant** [comment]
Yes, what has been described so far is perfectly normal and if you're binding an event to a method from the kivy api which returns a truthy object (and you don't want to eat the event), @bionoid's solution is the way to go.

Besides the better doc section @kived is mentioning, ModalView's `open` and `dismiss` return the widget, which is not documented. I'm also wondering why must the instance be returned?

**Comment by rafalo1333** [comment]
@dessant +1, on_dismiss fires the dialog instance, so I dont know why on_dismiss must also return the value at all?

**Comment by dessant** [comment]
I'm talking about the `dismiss` method which you've used in your example, not the `on_dismiss` event.

**Comment by rafalo1333** [comment]
It was a typo, sorry, I was talking about dismiss() too.

**Comment by kived** [comment]
@dessant I'm not sure why those methods return self. Generally this is done to create a fluent interface via method chaining, i.e. `my_object.do_this().do_that().do_another_thing().done()`. Not sure why this would be particularly handy for those methods though.

**Comment by dessant** [comment]
@kived, yeah, I don't see the advantage in this case either, I'll make a pr.

**Comment by rafalo1333** [comment]
Great, I believe dismiss/open methods just should do what they name, i.e close or open the modalview instance, without returning the value.

Thanks guys for help with the issue and thanks @dessant for PR :)

It seems we can close the issue?

**Comment by dessant** [comment]
@rafalo1333, we are repurposing it for the event doc. :wink:

**Comment by Zen-CODE** [comment]
Just wanted to point out this could be a breaking change. It's quite foreseeable that the 'on_dismiss' event could be used and bound to a method that uses the returned value. e.g.

    def dismissed(popup_ret):
        popup_ret.my_object.do_stuff()

    popup = Popup(title='Test', content=Label(text='Blah'))
    popup.object = MyObject()
    popup.bind(on_dismiss=dismissed)
    popup.open()

Personally, I think it's a dangerous change. I don't disagree that on_dismiss probably should not return anything, but it may well break people's code. And personally, I think it's a feature rather than a bug. e.g. the above code would allow you to easily pass info around without any other references...

Anyway, just my 2c. Feel free to ignore... :-)

**Comment by Zen-CODE** [comment]
To second that, look at the last example here before the API. It clearly demonstrates that 'on_dismiss' should return the instance.

https://kivy.org/docs/api-kivy.uix.popup.html

    def my_callback(instance):
        print('Popup', instance, 'is being dismissed but is prevented!')
        return True

    popup = Popup(content=Label(text='Hello world'))
    popup.bind(on_dismiss=my_callback)
    popup.open()

This is breaking documented behaviour...

**Comment by dessant** [comment]
@Zen-CODE, there may be a mixup between dismiss and on_dismiss, the change was made for the dismiss method, as we couldn't identify a good reason for the instance return, on_dismiss was not touched. https://github.com/kivy/kivy/commit/c9736d4b0ff25dcb2ee3503a032c76dc0c12967a

**Comment by Zen-CODE** [comment]
Great, thanks.
